cpp/tensorrt_llm/kernels/moeUtilOp.cu

```diff
@@ -0,0 +1,881 @@
+            expert_first_token_offset[out_expert_id] = local_expert_first_token_offset[expert_id];
+        }
+    }
+}
+
+template <int BLOCK_SIZE, int EXPERTS_PER_TOKEN, int LOG2_NUM_EXPERTS>
+bool fusedBuildExpertMapsSortFirstTokenDispatch(int const* token_selected_experts,
+    int* unpermuted_token_selected_experts, int* permuted_source_token_ids, int64_t* expert_first_token_offset,
+    int64_t const num_tokens, int const num_experts_per_node, int const experts_per_token, int const start_expert,
+    int const end_expert, cudaStream_t stream)
+{
+    TLLM_CHECK_WITH_INFO(num_experts_per_node == (end_expert - start_expert),
+        "num_experts_per_node must be equal to end_expert - start_expert");
+    int const threads = BLOCK_SIZE;
+    int const blocks = (num_tokens + threads - 1) / threads;
+    TLLM_CHECK_WITH_INFO(blocks == 1, "Current implementation requires single block");
+
+    using BlockRadixRank = cub::BlockRadixRank<BLOCK_SIZE, LOG2_NUM_EXPERTS, false>;
+    size_t shared_size = sizeof(typename BlockRadixRank::TempStorage);
+
+    cudaLaunchConfig_t config;
+    config.gridDim = blocks;
+    config.blockDim = threads;
+    config.dynamicSmemBytes = shared_size;
+    config.stream = stream;
+    cudaLaunchAttribute attrs[1];
+    attrs[0].id = cudaLaunchAttributeProgrammaticStreamSerialization;
+    attrs[0].val.programmaticStreamSerializationAllowed = tensorrt_llm::common::getEnvEnablePDL();
+    config.numAttrs = 1;
+    config.attrs = attrs;
+
+    auto kernel = &fusedBuildExpertMapsSortFirstTokenKernel<BLOCK_SIZE, EXPERTS_PER_TOKEN, LOG2_NUM_EXPERTS>;
+
+    int device = 0;
+    int max_smem_per_block = 0;
+    check_cuda_error(cudaGetDevice(&device));
+    check_cuda_error(cudaDeviceGetAttribute(&max_smem_per_block, cudaDevAttrMaxSharedMemoryPerBlockOptin, device));
+    if (shared_size >= static_cast<size_t>(max_smem_per_block))
+    {
+        // This should mean that
+        // cudaFuncSetAttribute(cutlass::Kernel<GemmKernel>, cudaFuncAttributeMaxDynamicSharedMemorySize, smem_size)
+        // wouldn't work.
+        return false;
+    }
+
+    check_cuda_error(cudaFuncSetAttribute(kernel, cudaFuncAttributeMaxDynamicSharedMemorySize, shared_size));
+    check_cuda_error(cudaLaunchKernelEx(&config, kernel, token_selected_experts, unpermuted_token_selected_experts,
+        permuted_source_token_ids, expert_first_token_offset, num_tokens, experts_per_token, start_expert, end_expert,
+        num_experts_per_node));
+
+    return true;
+}
+
+template <int EXPERTS_PER_TOKEN, int LOG2_NUM_EXPERTS>
+bool fusedBuildExpertMapsSortFirstTokenBlockSize(int const* token_selected_experts,
+    int* unpermuted_token_selected_experts, int* permuted_source_token_ids, int64_t* expert_first_token_offset,
+    int64_t const num_tokens, int const num_experts_per_node, int const experts_per_token, int const start_expert,
+    int const end_expert, cudaStream_t stream)
+{
+    int const block_size = num_tokens;
+    if (num_tokens > 256)
```

Comment:
This is suprising. 

Is there a fallback path to support num tokens > 256?

Reply:
Yes. The logic is as follows: 
num tokens <= 256: call fusedBuildExpertMapsSortFirstToken
num tokens > 256, call buildExpertMaps->generateTokenPermutation